In [1]:
import sys
sys.path.append('../src')
# Import necessary modules
from eda import EDA

In [2]:

# Initialize EDA class
eda = EDA(train_path='../data/raw/train.csv', store_path='../data/raw/store.csv')


In [3]:

# Load and merge datasets
merged_df = eda.load_and_merge_data()
print("Data merged successfully.")

2025-01-03 14:00:50,724 - EDA - INFO - Loading train and store datasets...
c:\Users\bksam\Rossmann-Sales-Forecasting\notebooks\../src\eda.py:17: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  train_df = pd.read_csv(self.train_path)
2025-01-03 14:00:51,211 - EDA - INFO - Datasets loaded successfully. Merging them...
2025-01-03 14:00:51,392 - EDA - INFO - Datasets merged successfully.


Data merged successfully.


Loaded and merged our train data and store data to perform Data exploration

In [5]:
# Explore Data
head, info, stats = eda.explore_data()

2025-01-03 14:01:03,826 - EDA - INFO - Exploring dataset...
2025-01-03 14:01:03,833 - EDA - INFO - First 5 Rows:
   Store  DayOfWeek        Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5  2015-07-31   5263        555     1      1            0   
1      2          5  2015-07-31   6064        625     1      1            0   
2      3          5  2015-07-31   8314        821     1      1            0   
3      4          5  2015-07-31  13995       1498     1      1            0   
4      5          5  2015-07-31   4822        559     1      1            0   

   SchoolHoliday StoreType Assortment  CompetitionDistance  \
0              1         c          a               1270.0   
1              1         a          a                570.0   
2              1         a          a              14130.0   
3              1         c          c                620.0   
4              1         a          a              29910.0   

   CompetitionOpenSinceMonth  Competiti

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 18 columns):
 #   Column                     Non-Null Count    Dtype  
---  ------                     --------------    -----  
 0   Store                      1017209 non-null  int64  
 1   DayOfWeek                  1017209 non-null  int64  
 2   Date                       1017209 non-null  object 
 3   Sales                      1017209 non-null  int64  
 4   Customers                  1017209 non-null  int64  
 5   Open                       1017209 non-null  int64  
 6   Promo                      1017209 non-null  int64  
 7   StateHoliday               1017209 non-null  object 
 8   SchoolHoliday              1017209 non-null  int64  
 9   StoreType                  1017209 non-null  object 
 10  Assortment                 1017209 non-null  object 
 11  CompetitionDistance        1014567 non-null  float64
 12  CompetitionOpenSinceMonth  693861 non-null   float64
 13  CompetitionO

2025-01-03 14:01:04,861 - EDA - INFO - Basic Statistics:
              Store     DayOfWeek         Sales     Customers          Open  \
count  1.017209e+06  1.017209e+06  1.017209e+06  1.017209e+06  1.017209e+06   
mean   5.584297e+02  3.998341e+00  5.773819e+03  6.331459e+02  8.301067e-01   
std    3.219087e+02  1.997391e+00  3.849926e+03  4.644117e+02  3.755392e-01   
min    1.000000e+00  1.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    2.800000e+02  2.000000e+00  3.727000e+03  4.050000e+02  1.000000e+00   
50%    5.580000e+02  4.000000e+00  5.744000e+03  6.090000e+02  1.000000e+00   
75%    8.380000e+02  6.000000e+00  7.856000e+03  8.370000e+02  1.000000e+00   
max    1.115000e+03  7.000000e+00  4.155100e+04  7.388000e+03  1.000000e+00   

              Promo  SchoolHoliday  CompetitionDistance  \
count  1.017209e+06   1.017209e+06         1.014567e+06   
mean   3.815145e-01   1.786467e-01         5.430086e+03   
std    4.857586e-01   3.830564e-01         7.715324e+0

### Data Exploration
In this step, we explored the dataset by checking its basic properties and summarizing the key statistics:

1. **First 5 Rows**:
   - Displayed the first 5 rows to get an initial view of the dataset's structure and the type of data it holds.
   
2. **Data Information**:
   - Provided an overview of the dataset, including the number of entries (1,017,209) and columns (18).
   - Listed the column names, data types, and the number of non-null values in each column.
   - Notably, some columns like `CompetitionDistance`, `CompetitionOpenSinceMonth`, and `Promo2SinceWeek` contain missing values.

3. **Basic Statistics**:
   - Displayed summary statistics for numerical columns, including mean, standard deviation, min, max, and quartiles.
   - For instance, `Sales` has a mean of 5,773.8 with a max value of 41,551, while `CompetitionDistance` has a mean of 5,430.1 with a large range.


In [7]:
# Check Missing Values
missing_values_df = eda.check_missing_values()

2025-01-03 14:23:12,603 - EDA - INFO - Checking for missing values in the dataset...
2025-01-03 14:23:12,978 - EDA - INFO - Missing values calculated successfully.
2025-01-03 14:23:12,989 - EDA - INFO - Missing Values Summary:
                           Missing Values  Percentage
Promo2SinceWeek                    508031   49.943620
PromoInterval                      508031   49.943620
Promo2SinceYear                    508031   49.943620
CompetitionOpenSinceYear           323348   31.787764
CompetitionOpenSinceMonth          323348   31.787764
CompetitionDistance                  2642    0.259730


### Missing Data
Next, we focused on checking and understanding the missing data in the dataset:

1. **Checking for Missing Values**:
   - We identified that some columns have missing values, specifically `Promo2SinceWeek`, `PromoInterval`, `Promo2SinceYear`, and others.
   
2. **Missing Values Summary**:
   - A summary was provided listing the columns with missing values along with the number of missing entries and the percentage of missing data.
   - Key columns with missing data include:
     - `Promo2SinceWeek`, `PromoInterval`, and `Promo2SinceYear` (all have around 50% missing values).
     - `CompetitionOpenSinceYear` and `CompetitionOpenSinceMonth` (approximately 31.7% missing values).
     - `CompetitionDistance` (only a small fraction of missing values—0.26%).

### Conclusion and Next Steps

**Conclusion**:
- The dataset contains a significant amount of missing values, particularly in the columns related to promotions (`Promo2SinceWeek`, `PromoInterval`, `Promo2SinceYear`), and competition-related features (`CompetitionOpenSinceYear`, `CompetitionOpenSinceMonth`).
- While some columns, like `CompetitionDistance`, have minimal missing data, others, like `PromoInterval` and `Promo2SinceWeek`, have a considerable percentage of missing values (~50%).

**Next Steps**
1. **Clean data**
2. **Detect outliners**
3. **Handle outliners using appropraite methods**